README.md
# Rackspace Database as a Service Node.js Bindings

Initial Project to open up the Rackspace DBaaS to Node.js through there API.

Based upon https://github.com/nodejitsu/node-cloudservers as much of the api structure is the same

### Installing clouddatabase
<pre>
  npm install clouddatabase
</pre>


### Getting Started
Before we can do anything with clouddatabase, we have to create a client with valid credentials. Clouddatabase will authenticate for you automatically:
<pre>
  var clouddatabase = require('clouddatabase');
  var config = {
    auth : {
      username: 'your-username',
      apiKey: 'your-api-key'
    }
  };
  var client = clouddatabase.createClient(config);
</pre>


CHANGELOG.md
## ChangeLog for: node-clouddatabase

### Version 0.0.2 - 18/4/2012
- Rewritten and added all api features to module

### Version 0.0.1 - 14/4/2012
- Initial Api clone from Cloudsevers

test/helpers.js
/*
 * helpers.js: Test helpers for node-clouddatabase
 *
 * Originally written by Nodejitsu Modified for cloud Database
 * (C) 2012 Mark Willis.
 * MIT LICENSE
 *
 */

var assert = require('assert'),
    fs = require('fs'),
    path = require('path'),
    util = require('util'),
    vows = require('vows'),
    clouddatabase = require('../lib/clouddatabase');

var helpers = exports,
    testConfig,
    client;

helpers.createClient = function () {
  if (!testConfig) {
    helpers.loadConfig();
  }

  if (!client) {
    client = clouddatabase.createClient(testConfig);
  }

  return client;
};

helpers.loadConfig = function () {
  try {
    var configFile = path.join(__dirname, 'fixtures', 'test-config.json'),
        stats = fs.statSync(configFile),
        config = JSON.parse(fs.readFileSync(configFile).toString());

    if (config.auth.username === 'test-username'
        || config.auth.apiKey === 'test-apiKey') {
      util.puts('Config file test/fixtures/test-config.json must be created with valid data before running tests');
      process.exit(0);
    }

    testConfig = config;
    return config;
  }
  catch (ex) {
    util.puts('Config file test/fixtures/test-config.json must be created with valid data before running tests');
    process.exit(0);
  }
};

helpers.assertServer = function (server) {
  assert.instanceOf(server, clouddatabase.Server);
  assert.isNotNull(server.id);
  assert.isNotNull(server.name);
};

helpers.assertServerDetails = function (server) {
  helpers.assertServer(server);
  assert.isNotNull(server.progess);
  assert.isNotNull(server.imageId);
  assert.isNotNull(server.flavorId);
  assert.isNotNull(server.status);
  assert.isNotNull(server.hostId);
  assert.isNotNull(server.addresses);
};

helpers.assertImage = function (image) {
  assert.instanceOf(image, clouddatabase.Image);
  assert.isNotNull(image.id);
  assert.isNotNull(image.name);
};

helpers.assertImageDetails = function (image) {
  helpers.assertImage(image);
  assert.isNotNull(image.updated);
  assert.isNotNull(image.created);
  assert.isNotNull(image.status);
};

helpers.assertFlavor = function (flavor) {
  assert.instanceOf(flavor, clouddatabase.Flavor);
  assert.isNotNull(flavor.id);
  assert.isNotNull(flavor.name);
};

helpers.assertFlavorDetails = function (flavor) {
  helpers.assertFlavor(flavor);
  assert.isNotNull(flavor.ram);
  assert.isNotNull(flavor.disk);
};

test/flavor-test.js
/*
 * flavor-test.js: Tests for rackspace clouddatabase flavor requests
 *
 * (C) 2010 Nodejitsu Inc.
 * MIT LICENSE
 *
 */

var fs = require('fs'),
    path = require('path'),
    vows = require('vows'),
    assert = require('assert'),
    clouddatabase = require('../lib/clouddatabase'),
    helpers = require('./helpers');

var testData = {},
    testContext = {},
    client = helpers.createClient();

vows.describe('node-clouddatabase/flavors').addBatch({
  "The node-clouddatabase client": {
    "the getFlavors() method": {
      "with no details": {
        topic: function () {
          client.getFlavors(this.callback);
        },
        "should return the list of flavors": function (err, flavors) {
          testContext.flavors = flavors;
          flavors.forEach(function (flavor) {
            helpers.assertFlavor(flavor);
          });
        }
      },
      "with details": {
        topic: function () {
          client.getFlavors(true, this.callback);
        },
        "should return the list of flavors": function (err, flavors) {
          flavors.forEach(function (flavor) {
            helpers.assertFlavorDetails(flavor);
          });
        }
      }
    }
  }
}).export(module);

test/authentication-test.js
/*
 * authentication-test.js: Tests for rackspace clouddatabase authentication
 *
 * Originally written by Nodejitsu Modified for cloud Database
 * (C) 2012 Mark Willis.
 * MIT LICENSE
 *
 */

var fs = require('fs'),
    path = require('path'),
    vows = require('vows'),
    assert = require('assert'),
    clouddatabase = require('../lib/clouddatabase'),
    helpers = require('./helpers');

var testData = {},
    client = helpers.createClient();

vows.describe('node-clouddatabase/authentication').addBatch({
  "The node-clouddatabase client": {
    "should have core methods defined": function () {
      assert.isObject(client.config.auth);
      assert.include(client.config.auth, 'username');
      assert.include(client.config.auth, 'apiKey');

      assert.isFunction(client.setAuth);
      assert.isFunction(client.getInstance);
    },
    "with a valid username and api key": {
      topic: function () {
        client.setAuth(this.callback);
      },
      "should respond with 200 and appropriate headers": function (err, res) {
        assert.equal(res.statusCode, 200);
        var resBody = JSON.parse(res.body);
        assert.isObject(resBody);
      },
      "should update the config with appropriate urls": function (err, res) {
        var config = client.config;
        var resBody = JSON.parse(res.body);
        assert.isObject(resBody);
      }
    },
    "with an invalid username and api key": {
      topic: function () {
        var badClient = clouddatabase.createClient({
          "auth": {
            "username": "fake",
            "apiKey": "data"
          }
        });

        badClient.setAuth(this.callback);
      },
      "should respond with 401": function (err, res) {
        assert.equal(err, "Username or api key is invalid");
      }
    }
  }
}).export(module);

trialRun.js
var clouddatabase = require('./lib/clouddatabase'),
	helpers = require('./test/helpers');

var client = helpers.createClient();


console.log("Setup Auth");

client.setAuth(function(err) {

	console.log("Get List of Instances");
	client.getInstances(/*true,*/function(err, instances) {
		instances.forEach(function(item) {
			console.log("name: "+item.name+ ": " + item.status);
		});
		if(instances.length) {
			var item = instances.pop();
			if(item.status == "ACTIVE") {
				processDB(item); // run all commands == shutdown
			} else {
				console.log("Wait for server finish");
				item.setWait({ status: 'ACTIVE' }, 5000, function () {
					console.log("Should be active");
				});
			}
		} else {
			console.log("No Servers -- Start One");

			console.log("Get Flavors");
			client.getFlavors(function(err,flavors) {
				var useFlavor = flavors.pop();
				//console.log(useFlavor.links[0].href);
				client.addDatabase(useFlavor.links[0].href, 1, "nodeTest", "db1", done);

			});


		}
	});

});



function done(err,body) {
	console.log(err);
	console.log(body);
	console.log("done");
}

function processDB(item) {
	console.log("ID: "+item.id);
	console.log("Name: "+item.name);
	console.log("Status: "+item.status);
	item.getDetails(function(err, items /* not used */) { // got all instance details

//			console.log(item);

/*
		console.log("Delete Instance")
		item.deleteServer(function(err) { // delete the instace we just made
			if(err) {
				console.log("Error on delete: "+ err);
			}
		}); */

		console.log("List Databases");
		item.getDatabases(function(err, databases) {
			console.log(databases);
		});

/*		console.log("Add New Database");
		var options = [{
			name: "NewDB2"
		}];

		item.addDatabase(options, function(err) {
			if(err) {
				console.log("Error on add: "+ err);
			}
		}); */

/*		console.log("Delete Databases");
		item.deleteDatabase("NewDB2", function(err) {
			if(err) {
				console.log("Error on delete: "+ err);
			}
		}); */


		console.log("List Users");
		item.getUsers(function(err, users) {
			console.log(users);
		});

/*		console.log("Add New User");
		var options = [{
			name: "mark",
			password: "mark",
			database: "db1"
		}];

		item.addUser(options, function(err) {
			if(err) {
				console.log("Error on add: "+ err);
			}
		});
*/

/*		console.log("Delete User");
		item.deleteUser("mark", function(err) {
			if(err) {
				console.log("Error on delete: "+ err);
			}
		});
*/

/*		console.log("Resize Instance - Change Flavor");

		item.resizeFlavor({flavorRef: 2}, function(err) {
			if(err) {
				console.log("Error on resize: "+ err);
			}
			console.log("Waiting for resize");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		}); */


/*
		console.log("Resize Instance - Change Flavor");

		item.resizeFlavor({flavorRef: 2}, function(err) {
			if(err) {
				console.log("Error on resize: "+ err);
			}
			console.log("Waiting for resize");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		}); */

/*
		console.log("Resize Instance - Volume");

		item.resizeVolume({volume: { size: 3 } }, function(err) {
			if(err) {
				console.log("Error on volume resize: "+ err);
			}
			console.log("Waiting for resize");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		});
*/

/*
		console.log("Restart Instance");

		item.restart(function(err) {
			if(err) {
				console.log("Error on restart: "+ err);
			}
			console.log("Waiting for restart");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		});
*/

/*
		console.log("Check Root User Enabled");

		item.checkRootUser(function(err, body) {
			if(err) {
				console.log("Error on check root user: "+ err);
			}
			console.log(body);

		});
*/

/*
		console.log("Enable Root User");

		item.enableRootUser(function(err, body) {
			if(err) {
				console.log("Error on enable root user: "+ err);
			}
			console.log(body);

		});
*/

	});
}

examples/allFeatures.js
var clouddatabase = require('..//lib/clouddatabase'),
	helpers = require('../test/helpers');

var client = helpers.createClient();


console.log("Setup Auth");

client.setAuth(function(err) {

	console.log("Get List of Instances");
	client.getInstances(/*true,*/function(err, instances) {
		instances.forEach(function(item) {
			console.log("name: "+item.name+ ": " + item.status);
		});
		if(instances.length) {
			var item = instances.pop();
			if(item.status == "ACTIVE") {
				processDB(item); // run all commands == shutdown
			} else {
				console.log("Wait for server finish");
				item.setWait({ status: 'ACTIVE' }, 5000, function () {
					console.log("Should be active");
				});
			}
		} else {
			console.log("No Servers -- Start One");

			console.log("Get Flavors");
			client.getFlavors(function(err,flavors) {
				var useFlavor = flavors.pop();
				//console.log(useFlavor.links[0].href);
				client.addDatabase(useFlavor.links[0].href, 1, "nodeTest", "db1", done);

			});


		}
	});

});



function done(err,body) {
	console.log(err);
	console.log(body);
	console.log("done");
}

function processDB(item) {
	console.log("ID: "+item.id);
	console.log("Name: "+item.name);
	console.log("Status: "+item.status);

	item.getDetails(function(err, items /* not used */) { // got all instance details

		console.log("Hostname: "+item.hostname);

//			console.log(item);

/*
		console.log("Delete Instance")
		item.deleteServer(function(err) { // delete the instace we just made
			if(err) {
				console.log("Error on delete: "+ err);
			}
		}); */

		console.log("List Databases");
		item.getDatabases(function(err, databases) {
			console.log(databases);
		});

/*		console.log("Add New Database");
		var options = [{
			name: "NewDB2"
		}];

		item.addDatabase(options, function(err) {
			if(err) {
				console.log("Error on add: "+ err);
			}
		}); */

/*		console.log("Delete Databases");
		item.deleteDatabase("NewDB2", function(err) {
			if(err) {
				console.log("Error on delete: "+ err);
			}
		}); */


		console.log("List Users");
		item.getUsers(function(err, users) {
			console.log(users);
		});

/*		console.log("Add New User");
		var options = [
		{
			name: "mark",
			password: "mark",
			database: "db1"
		},
		{
			name: "mark2",
			password: "mark2",
			databases: [ {
				name: "db1"
			}]
		},

		];

		item.addUser(options, function(err) {
			if(err) {
				console.log("Error on add: "+ err);
			}
		});*/


/*		console.log("Delete User");
		item.deleteUser("mark", function(err) {
			if(err) {
				console.log("Error on delete: "+ err);
			}
		});
*/

/*		console.log("Resize Instance - Change Flavor");

		item.resizeFlavor({flavorRef: 2}, function(err) {
			if(err) {
				console.log("Error on resize: "+ err);
			}
			console.log("Waiting for resize");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		}); */


/*
		console.log("Resize Instance - Change Flavor");

		item.resizeFlavor({flavorRef: 2}, function(err) {
			if(err) {
				console.log("Error on resize: "+ err);
			}
			console.log("Waiting for resize");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		}); */

/*
		console.log("Resize Instance - Volume");

		item.resizeVolume({volume: { size: 3 } }, function(err) {
			if(err) {
				console.log("Error on volume resize: "+ err);
			}
			console.log("Waiting for resize");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		});
*/

/*
		console.log("Restart Instance");

		item.restart(function(err) {
			if(err) {
				console.log("Error on restart: "+ err);
			}
			console.log("Waiting for restart");
			item.setWait({ status: 'ACTIVE' }, 5000, function () {
				console.log("Should be active");
			});

		});
*/

/*
		console.log("Check Root User Enabled");

		item.checkRootUser(function(err, body) {
			if(err) {
				console.log("Error on check root user: "+ err);
			}
			console.log(body);

		});
*/

/*
		console.log("Enable Root User");

		item.enableRootUser(function(err, body) {
			if(err) {
				console.log("Error on enable root user: "+ err);
			}
			console.log(body);

		});
*/

	});
}

lib/clouddatabase/core.js
/*
 * core.js: Core functions for accessing rackspace cloud servers
 *
 * Originally written by Nodejitsu Modified for cloud Database
 * (C) 2012 Mark Willis.
 * MIT LICENSE
 *
 */

var http = require('http'),
    url = require('url'),
    request = require('request'),
    clouddatabase = require('../clouddatabase'),
    utils = require('./utils');

var defaultAuthUrl = 'lon.identity.api.rackspacecloud.com/',
    serverUrl = 'lon.databases.api.rackspacecloud.com/v1.0/';

exports.createClient = function (options) {
  return new Client(options);
};

var Client = exports.Client = function (options) {
  if (!options.auth) throw new Error ('options.auth is required to create Config');

  this.config = options;
  this.config.serverUrl = 'https://' + (this.config.serverUrl || serverUrl);
  this.authorized = false;
};

//
// Authenticates node-clouddatabase with the specified options:
// { username: "your-username", apiKey: "your-secret-key" }
//
Client.prototype.setAuth = function (callback) {
  var self = this;
  var authUrl = this.config.auth.host || defaultAuthUrl;

  var body = JSON.stringify({
      "credentials": {
        'username': this.config.auth.username,
        'key': this.config.auth.apiKey
      }
    });

  var authOptions = {
    uri: 'https://' + authUrl + 'v1.1/auth',
    headers: {
      'HOST': authUrl,
      'Content-Type' : 'application/json',
      'Accept' : 'application/json'
    },
    body: body
  };

  request.post(authOptions, function (err, res, body) {
    if (err) return callback(err);

    var jsonBody = JSON.parse(body);
    if(jsonBody.unauthorized) {
      callback(jsonBody.unauthorized.message);
      return;
    }

    self.authorized = true;
    self.config.serverUrl = self.config.serverUrl+self.config.auth.accountNumber;//jsonBody.auth.serviceCatalog.cloudServers[0].publicURL;
    self.config.authToken = jsonBody.auth.token.id;

    self.serverUrl = self.config.serverUrl;

    callback(null, res);
  });
};

//
// Gets all flavors (i.e. size configurations) for the authenticated username / apikey.
// Parameters: details? callback
//
Client.prototype.getFlavors = function () {
  var self = this, details = false, callback;

  if (typeof arguments[0] === 'function') {
    callback = arguments[0];
  }
  else {
    details = arguments[0];
    callback = arguments[1];
  }

  var flavorPath = details ? '/flavors/detail.json' : '/flavors.json';
  utils.rackspace([flavorPath], this, callback, function (body) {
    var flavorInfo = JSON.parse(body), results = [];
    flavorInfo.flavors.forEach(function (info) {
      results.push(new (clouddatabase.Flavor)(self, info));
    });

    callback(null, results);
  });
};

// create Database Server
Client.prototype.addDatabase = function (flavorId, dbSize, instanceName, databaseName, callback) {
  var options = {
    "instance": {
      "databases": [
          {
            "name": databaseName
          }
        ]
      ,
      "flavorRef": flavorId,
      "name": instanceName,
      "volume": {
        "size": dbSize.toString()
      }
    }
  }

  var instancePath = "/instances";

  var createOptions = {
    method: 'POST',
    uri: [instancePath],
    client: this,
    body: options
  };

  utils.rackspace(createOptions, callback, function (body) {
    callback(null, body);
  });

}


//
// Gets all instances of the any databases currently setup
// Parameters: details (boolean) , callback
//
Client.prototype.getInstances = function () {
  var self = this, details = false, callback;

  if (typeof arguments[0] === 'function') {
    callback = arguments[0];
  }
  else {
    details = arguments[0];
    callback = arguments[1];
  }

  var flavorPath = details ? "/instances/detail" : "/instances";


  utils.rackspace([flavorPath], this, callback, function (body) {

    var instanceInfo = JSON.parse(body), results = [];
    instanceInfo.instances.forEach(function (info) {

      results.push(new (clouddatabase.Database)(self, info));
    });

    callback(null, results);
  });
}

//
// Gets details of single database
// Parameters: instanceId, callback
//
Client.prototype.getInstance = function (instanceId, callback) {
  var self = this;

  var flavorPath = "/instances/"+instanceId;


  utils.rackspace([flavorPath], this, callback, function (body) {
    var instanceInfo = JSON.parse(body), results = [];
    callback(null, instanceInfo);
  });

};


lib/clouddatabase/flavor.js
/*
 * Flavor.js: Instance of a single rackspace cloudserver flavor
 *
 * (C) 2010 Nodejitsu Inc.
 * MIT LICENSE
 *
 */

var Flavor = function (client, details) {
  if (!details) throw new Error("Flavor must be constructed with at-least basic details.")
  
  this.client = client;
  this._setProperties(details);
};

Flavor.prototype = {
  //
  // Gets details for this instance
  //
  getDetails: function (callback) {
    var self = this;
    this.client.getFlavor(this.id, function (err, flavor) {
      if (err) return callback(err);

      self._setProperties(flavor);
      callback(null, self);
    });
  },
  
  //
  // Sets the properties for this instance
  // Parameters: details
  //
  _setProperties: function (details) {
    this.id = details.id;
    this.name = details.name;
    this.ram = details.ram;
    this.disk = details.disk;
    this.links = details.links;
  }
};

exports.Flavor = Flavor;

lib/clouddatabase/database.js
/**
	Rackspace database api calls.
**/

var utils = require("./utils");

var Database = function (client, details) {
  if (!details) throw new Error("CloudDatabase must be constructed with at least basic details.")

  	this.client = client;
	this._setProperties(details);

};

Database.prototype = {
	//
	// Helper method for performing 'Server Actions' to /instances/:id/action
	// e.g. Reboot, Rebuild, Resize, Confirm Resize, Revert Resize
	//
	databaseAction: function (id, body, client, callback) {
	  var actionOptions = {
	    method: 'POST',
	    client: client,
	    uri: '/instances/'+id+'/action',
	    body: body
	  };

	  utils.rackspace(actionOptions, callback, function (body, res) {
	    callback(null, res);
	  });
	},

	//
	// Sets the properties for this instance
	// Parameters: details
	//
	_setProperties: function (details) {
		// Set core properties
		this.id = details.id;
		this.links = details.links;
		this.status = details.status;
		this.name = details.name;
		this.flavor = details.flavor;
		this.created = details.created;
		this.updated = details.updated;
		this.hostname = details.hostname;
		this.volume = details.volume;
		this.rootEnabled = details.rootEnabled;


	},

	getDetails: function(callback) {
		var self = this;
		this.client.getInstance(this.id, function(err, info) {
			self._setProperties(info.instance);
			callback(null, self);
		})
	},

	getDatabases: function(callback) {
	  var actionOptions = {
	    method: 'GET',
	    client: this.client,
	    uri: '/instances/'+this.id+'/databases'
	  };


	  utils.rackspace(actionOptions, callback, function (body) {
	    callback(null, body);
	  });
	},

	addDatabase: function(options, callback) {
	  var actionOptions = {
	    method: 'POST',
	    client: this.client,
	    uri: '/instances/'+this.id+'/databases',
	    body: {databases: options}
	  };

	  utils.rackspace(actionOptions, callback, function (body) {
	    callback(null, body);
	  });
	},

	deleteDatabase: function(databaseName, callback) {
	  var actionOptions = {
	    method: 'DELETE',
	    client: this.client,
	    uri: '/instances/'+this.id+'/databases/'+databaseName
	  };

	  utils.rackspace(actionOptions, callback, function (body) {
	    callback(null, body);
	  });
	},

	getUsers: function(callback) {
	  var actionOptions = {
	    method: 'GET',
	    client: this.client,
	    uri: '/instances/'+this.id+'/users'
	  };


	  utils.rackspace(actionOptions, callback, function (body) {
	    callback(null, body);
	  });
	},

	addUser: function(options, callback) {
	  var actionOptions = {
	    method: 'POST',
	    client: this.client,
	    uri: '/instances/'+this.id+'/users',
	    body: {users: options}
	  };

	  utils.rackspace(actionOptions, callback, function (body) {
	    callback(null, body);
	  });
	},

	deleteUser: function(databaseName, callback) {
	  var actionOptions = {
	    method: 'DELETE',
	    client: this.client,
	    uri: '/instances/'+this.id+'/users/'+databaseName
	  };

	  utils.rackspace(actionOptions, callback, function (body) {
	    callback(null, body);
	  });
	},

	deleteServer: function(callback) {
	  var actionOptions = {
	    method: 'DELETE',
	    client: this.client,
	    uri: '/instances/'+this.id,
	  };


	  utils.rackspace(actionOptions, callback, function (body) {
	    callback(null, body);
	  });

	},

	getVersion: function () {
		console.log("something");

	},

	resizeFlavor: function(options, callback) {
		var body = {resize: options};
		this.databaseAction(this.id, body, this.client, callback);
	},

	resizeVolume: function(options, callback) {
		options.volume.size = options.volume.size.toString();
		var body = {resize: options};
		this.databaseAction(this.id, body, this.client, callback);
	},

	restart: function(callback) {
		var body = {restart: {}};
		this.databaseAction(this.id, body, this.client, callback);
	},

	enableRootUser: function(callback) {
		var actionOptions = {
			method: 'POST',
			client: this.client,
			uri: '/instances/'+this.id+'/root',
		};


		utils.rackspace(actionOptions, callback, function (body) {
			callback(null, body);
		});
	},

	checkRootUser: function(callback) {
		var actionOptions = {
			method: 'GET',
			client: this.client,
			uri: '/instances/'+this.id+'/root',
		};


		utils.rackspace(actionOptions, callback, function (body) {
			callback(null, body);
		});
	},


	//
	// Continually polls Rackspace CloudServers and checks the
	// results against the attributes parameter. When the attributes
	// match the callback will be fired.
	// Parameters: attributes callback
	//
	setWait: function (attributes, interval, callback) {
	var self = this;
	var equalCheckId = setInterval(function () {
	  self.getDetails(function (err, server) {
	    if (err) return; // Ignore errors

	    var equal = true, keys = Object.keys(attributes);
	    for (index in keys) {
	      if (attributes[keys[index]] !== server[keys[index]]) {
	        equal = false;
	        break;
	      }
	    }

	    if (equal) {
	      clearInterval(equalCheckId);
	      callback(null, self);
	    }
	  });
	}, interval);

	return equalCheckId;
	},

	//
	// Clears a previously setWait for this instance
	// Parameters: intervalId
	//
	clearWait: function (intervalId) {
	clearInterval(intervalId);
	},
};

exports.Database = Database;

lib/clouddatabase/utils.js
/*
 * core.js: Core functions for accessing rackspace cloud servers
 *
 * Originally written by Nodejitsu Modified for cloud Database
 * (C) 2012 Mark Willis.
 * MIT LICENSE
 *
 */

var spawn = require('child_process').spawn,
    exec = require('child_process').exec,
    request = require('request'),
    clouddatabase = require('../clouddatabase');

var utils = exports;

// Failure HTTP Response codes based
// off Rackspace clouddatabase specification.
var failCodes = {
  400: "Bad Request",
  401: "Unauthorized",
  403: "Resize not allowed",
  404: "Item not found",
  409: "Build in progress",
  413: "Over Limit",
  415: "Bad Media Type",
  500: "Fault",
  503: "Service Unavailable"
};

// Export the set of Failure Codes
utils.failCodes = failCodes;

// Success HTTP Response codes based
// off Rackspace clouddatabase specification.
var successCodes = {
  200: "OK",
  202: "Accepted",
  203: "Non-authoritative information",
  204: "No content",
};

// Export the set of Success Codes
utils.successCodes = successCodes;

//
// Core method that actually sends requests to Rackspace.
// This method is designed to be flexible w.r.t. arguments
// and continuation passing given the wide range of different
// requests required to fully implement the clouddatabase API.
//
// Continuations:
//   1. 'callback': The callback passed into every node-clouddatabase method
//   2. 'success':  A callback that will only be called on successful requests.
//                  This is used throughout node-clouddatabase to conditionally
//                  do post-request processing such as JSON parsing.
//
// Possible Arguments (1 & 2 are equivalent):
//   1. utils.rackspace('some-fully-qualified-url', client, callback, success)
//   2. utils.rackspace('GET', 'some-fully-qualified-url', client, callback, success)
//   3. utils.rackspace('DELETE', 'some-fully-qualified-url', client,  callback, success)
//   4. utils.rackspace({ method: 'POST', uri: 'some-url', body: { some: 'body'}, client: new clouddatabase.Client }, callback, success)
//
utils.rackspace = function () {
  var args = Array.prototype.slice.call(arguments),
      success = (typeof(args[args.length - 1]) === 'function') && args.pop(),
      callback = (typeof(args[args.length - 1]) === 'function') && args.pop(),
      uri, method, requestBody, client, headers = {};

  // Now that we've popped off the two callbacks
  // We can make decisions about other arguments
  if (args.length == 1) {
    method = args[0]['method'] || 'GET',
    uri = args[0]['uri'];
    requestBody = args[0]['body'];
    client = args[0]['client'];
  }
  else if (args.length === 2) {
    // If we got a string assume that it's the URI
    method = 'GET';
    uri = args[0];
    client = args[1];
  }
  else {
    method = args[0];
    uri = args[1];
    client = args[2];
  }

  //
  // Makes the raw request to Rackspace Cloudfiles
  //
  function makeRequest () {
    // Append the X-Auth-Token header for Rackspace authenticatio
    headers['X-Auth-Token'] = client.config.authToken;
    var serverOptions = {
      uri: uri.replace(/\.json$/,''),
      method: method,
      headers: headers
    };

    if (typeof requestBody !== 'undefined') {
      serverOptions.headers['Content-Type'] = 'application/json';
      serverOptions.body = JSON.stringify(requestBody);
    }
    else if (typeof requestBody !== 'undefined') {
      serverOptions.body = requestBody;
    }

//console.log(serverOptions);

    request(serverOptions, function (err, res, body) {
      if (err) {
        if (callback) callback(err);
        return;
      }

      var statusCode = res.statusCode.toString();

      if (Object.keys(failCodes).indexOf(statusCode) !== -1) {
        if (callback) callback(new Error('Rackspace Error (' + statusCode + '): ' + failCodes[statusCode]));
        return;
      }

      success(body, res);
    });
  }

  if (client.authorized) {
      uri = client.serverUrl+uri;
      makeRequest();
  }
  else {
    client.setAuth(function (err, res) {
      if (err) return callback(err);
//      uri = client.serverUrl.apply(client, [].concat(uri));
      uri = client.serverUrl+uri;
      makeRequest();
    });
  }
};

var contentTypeOptions = '-i -H "Content-Type: application/json" ';
var authTokenOptions = '-i -H "X-Auth-Token:{{auth-token}}" ';
var curlOptions = '-X "{{method}}" {{uri}}';

utils.rackspaceCurl = function (method, uri, client, callback, success) {
  function makeRequest() {
    var options = 'curl ', error = '', data = '';

    if (method === 'POST') {
      options += contentTypeOptions;
    }

    options += authTokenOptions.replace('{{auth-token}}', client.config.authToken);
    options += curlOptions.replace('{{method}}', method).replace('{{uri}}', uri);

    exec(options, function (error, stdout, stderr) {
      if (error) return callback(error);

      var statusCode = stdout.match(/HTTP\/1.1\s(\d+)/)[1];
      if (Object.keys(failCodes).indexOf(statusCode.toString()) !== -1) {
        return callback(new Error('Rackspace Error (' + statusCode + '): ' + failCodes[statusCode]));
      }

      callback(null, { statusCode: statusCode });
    });
  }
  if (client.authorized) {
      makeRequest();
  }
  else {
    client.setAuth(function (err, res) {
      if (err) return callback(err);
      uri = client.serverUrl() + uri;
      makeRequest();
    });
  }
};
